no `gnmi_set` found when running 'sudo make test'

```
*** Starting controller
c0 Timeout waiting for FAUCET to start. Log:
Traceback (most recent call last):
  File "/usr/local/bin/faucet", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.5/dist-packages/faucet/__main__.py", line 166, in main
    os.execvp(ryu_args[0], ryu_args)
  File "/usr/lib/python3.5/os.py", line 615, in execvp
    _execvpe(file, args)
  File "/usr/lib/python3.5/os.py", line 660, in _execvpe
    raise last_exc.with_traceback(tb)
  File "/usr/lib/python3.5/os.py", line 650, in _execvpe
    exec_func(fullname, *argrest)
FileNotFoundError: [Errno 2] No such file or directory

*** Starting 2 switches
s1 s2 ...
* Shutting down any agents listening on 10161
* Starting agent
* Waiting for agent to start up
10161/tcp:          
* Sending test configuration to agent
Traceback (most recent call last):
  File "./agenttest.py", line 399, in <module>
    main()
  File "./agenttest.py", line 394, in main
    exit_code = end_to_end_test()
  File "./agenttest.py", line 349, in end_to_end_test
    result = run(cmd, stdout=PIPE, check=True)
  File "/usr/lib/python3.5/subprocess.py", line 693, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib/python3.5/subprocess.py", line 947, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.5/subprocess.py", line 1551, in _execute_child
    raise child_exception_type(errno_num, err_msg)
FileNotFoundError: [Errno 2] No such file or directory: 'gnmi_set'
makefile:46: recipe for target 'test' failed
make: *** [test] Error 1
```

I pulled in the latest and ran `./dependencies.sh` and noticed this time it did this:
```
* Installing gnxi tools
package github.com/google/gnxi/gnmi_capabilities: cannot download, $GOPATH not set. For more details see: go help gopath
can't load package: package github.com/google/gnxi/gnmi_capabilities: cannot find package "github.com/google/gnxi/gnmi_capabilities" in any of:
	/usr/lib/go-1.6/src/github.com/google/gnxi/gnmi_capabilities (from $GOROOT)
	($GOPATH not set)
package github.com/google/gnxi/gnmi_get: cannot download, $GOPATH not set. For more details see: go help gopath
can't load package: package github.com/google/gnxi/gnmi_get: cannot find package "github.com/google/gnxi/gnmi_get" in any of:
	/usr/lib/go-1.6/src/github.com/google/gnxi/gnmi_get (from $GOROOT)
	($GOPATH not set)
package github.com/google/gnxi/gnmi_set: cannot download, $GOPATH not set. For more details see: go help gopath
can't load package: package github.com/google/gnxi/gnmi_set: cannot find package "github.com/google/gnxi/gnmi_set" in any of:
	/usr/lib/go-1.6/src/github.com/google/gnxi/gnmi_set (from $GOROOT)
	($GOPATH not set)
package github.com/google/gnxi/gnmi_target: cannot download, $GOPATH not set. For more details see: go help gopath
can't load package: package github.com/google/gnxi/gnmi_target: cannot find package "github.com/google/gnxi/gnmi_target" in any of:
	/usr/lib/go-1.6/src/github.com/google/gnxi/gnmi_target (from $GOROOT)
	($GOPATH not set)
* Installing FAUCET as root
* Done
```
It probably did that last time too, I just missed it. I added `export` in front of the `GOPATH` variable in the beginning of the script and it fixed the `GOPATH` issue, but resulted in this:
```
* Installing gnxi tools
package context: unrecognized import path "context" (import path does not begin with hostname)
package net/http/httptrace: unrecognized import path "net/http/httptrace" (import path does not begin with hostname)
../../../go/src/google.golang.org/grpc/balancer.go:22:2: cannot find package "context" in any of:
	/usr/lib/go-1.6/src/context (from $GOROOT)
	/home/clewis/go/src/context (from $GOPATH)
../../../go/src/golang.org/x/net/http2/not_go111.go:10:2: cannot find package "net/http/httptrace" in any of:
	/usr/lib/go-1.6/src/net/http/httptrace (from $GOROOT)
	/home/clewis/go/src/net/http/httptrace (from $GOPATH)
package context: unrecognized import path "context" (import path does not begin with hostname)
package net/http/httptrace: unrecognized import path "net/http/httptrace" (import path does not begin with hostname)
../../../go/src/google.golang.org/grpc/balancer.go:22:2: cannot find package "context" in any of:
	/usr/lib/go-1.6/src/context (from $GOROOT)
	/home/clewis/go/src/context (from $GOPATH)
../../../go/src/golang.org/x/net/http2/not_go111.go:10:2: cannot find package "net/http/httptrace" in any of:
	/usr/lib/go-1.6/src/net/http/httptrace (from $GOROOT)
	/home/clewis/go/src/net/http/httptrace (from $GOPATH)
package context: unrecognized import path "context" (import path does not begin with hostname)
package net/http/httptrace: unrecognized import path "net/http/httptrace" (import path does not begin with hostname)
../../../go/src/google.golang.org/grpc/balancer.go:22:2: cannot find package "context" in any of:
	/usr/lib/go-1.6/src/context (from $GOROOT)
	/home/clewis/go/src/context (from $GOPATH)
../../../go/src/golang.org/x/net/http2/not_go111.go:10:2: cannot find package "net/http/httptrace" in any of:
	/usr/lib/go-1.6/src/net/http/httptrace (from $GOROOT)
	/home/clewis/go/src/net/http/httptrace (from $GOPATH)
package context: unrecognized import path "context" (import path does not begin with hostname)
package net/http/httptrace: unrecognized import path "net/http/httptrace" (import path does not begin with hostname)
../../../go/src/google.golang.org/grpc/balancer.go:22:2: cannot find package "context" in any of:
	/usr/lib/go-1.6/src/context (from $GOROOT)
	/home/clewis/go/src/context (from $GOPATH)
../../../go/src/golang.org/x/net/http2/not_go111.go:10:2: cannot find package "net/http/httptrace" in any of:
	/usr/lib/go-1.6/src/net/http/httptrace (from $GOROOT)
	/home/clewis/go/src/net/http/httptrace (from $GOPATH)
* Installing FAUCET as root
* Done
```
For whatever reason it's angry about that context thing.

I updated to golang 1.12 and it fixed the gnxi/context issues.  There was however a few other things that cropped up.

1. had to run `./test-dependencies` with sudo for two of the pip installs
```
$ ./test-dependencies.sh 
* Installing testing (make {test,codecheck}) dependencies
Could not install packages due to an EnvironmentError: [Errno 13] Permission denied: '/usr/local/lib/python3.5/dist-packages/pycodestyle.py'
Consider using the `--user` option or check the permissions.

/tmp/tmp.5q4u4jOzga ~/github/faucetagent
Cloning into 'mininet'...
remote: Enumerating objects: 28, done.
remote: Counting objects: 100% (28/28), done.
remote: Compressing objects: 100% (21/21), done.
remote: Total 9645 (delta 9), reused 23 (delta 6), pack-reused 9617
Receiving objects: 100% (9645/9645), 3.00 MiB | 5.62 MiB/s, done.
Resolving deltas: 100% (6393/6393), done.
Checking connectivity... done.
Could not install packages due to an EnvironmentError: [Errno 13] Permission denied: '/usr/local/lib/python3.5/dist-packages/mininet-2.3.0d5.dist-info'
Consider using the `--user` option or check the permissions.
```

2. while the test got a lot farther this time, it still ultimately failed for me, and I'm not sure why:
```
$ sudo make test
GOPATH=/home/clewis/go PATH=/home/clewis/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin ./agenttest.py
* Generating certificates
* Generating fake CA cert
* Generating and signing fake server cert
* Generating fake client cert (self-signed)
* Starting network
*** Creating network
*** Adding controller
*** Adding hosts:
h1 h2 h3 h4 
*** Adding switches:
s1 s2 
*** Adding links:
(s1, h1) (s1, h2) (s1, s2) (s2, h3) (s2, h4) 
*** Configuring hosts
h1 h2 h3 h4 
*** Starting controller
c0 Timeout waiting for FAUCET to start. Log:
Traceback (most recent call last):
  File "/usr/local/bin/faucet", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.5/dist-packages/faucet/__main__.py", line 166, in main
    os.execvp(ryu_args[0], ryu_args)
  File "/usr/lib/python3.5/os.py", line 615, in execvp
    _execvpe(file, args)
  File "/usr/lib/python3.5/os.py", line 660, in _execvpe
    raise last_exc.with_traceback(tb)
  File "/usr/lib/python3.5/os.py", line 650, in _execvpe
    exec_func(fullname, *argrest)
FileNotFoundError: [Errno 2] No such file or directory

*** Starting 2 switches
s1 s2 ...
* Shutting down any agents listening on 10161
* Starting agent
* Waiting for agent to start up
10161/tcp:          
* Sending test configuration to agent
F0410 17:02:51.057616     799 gnmi_set.go:160] Set failed: rpc error: code = Unknown desc = Exception calling application: Command '['fuser', '-k', '-HUP', '9302/tcp']' returned non-zero exit status 1
Traceback (most recent call last):
  File "./agenttest.py", line 399, in <module>
    main()
  File "./agenttest.py", line 394, in main
    exit_code = end_to_end_test()
  File "./agenttest.py", line 349, in end_to_end_test
    result = run(cmd, stdout=PIPE, check=True)
  File "/usr/lib/python3.5/subprocess.py", line 708, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['gnmi_set', '-ca', 'testcerts/fakeca.crt', '-cert', 'testcerts/fakeclient.crt', '-key', 'testcerts/fakeclient.key', '-target_name', 'localhost', '-replace=/:\nvlans:\n  office:\n    vid: 100\n  guest:\n    vid: 200\n\ndps:\n  s1:\n    dp_id: 0x1\n    hardware: "Open vSwitch"\n    interfaces:\n      1:\n        name: "h1"\n        native_vlan: office\n      2:\n        name: "h2"\n        native_vlan: guest\n      3:\n        name: "link"\n        tagged_vlans: [office, guest]\n  s2:\n    dp_id: 0x2\n    hardware: "Open vSwitch"\n    interfaces:\n      1:\n        name: "h3"\n        native_vlan: office\n      2:\n        name: "h4"\n        native_vlan: guest\n      3:\n        name: "link"\n        tagged_vlans: [office, guest]\n']' returned non-zero exit status 1
makefile:46: recipe for target 'test' failed
make: *** [test] Error 1
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -13,8 +13,6 @@ before_install:
 install:
 - ./dependencies.sh
 - ./test-dependencies.sh
-- echo "Testing against latest FAUCET"
-- sudo pip3 -q install git+https://github.com/faucetsdn/faucet
 
 script:
 - make

--- a/dependencies.sh
+++ b/dependencies.sh
@@ -1,26 +1,12 @@
 #!/bin/bash
 
-GOPATH=${GOPATH:=$HOME/go}
+set -e
 
-echo "* Installing Python GRPC dependencies"
-  sudo apt install python3-pip python3-setuptools python3-wheel
-  pip3 -q install protobuf grpcio grpcio-tools requests
-  # In case we 'make clean; sudo make test'
-  sudo pip3 -q install --no-cache protobuf grpcio grpcio-tools requests
+PIP3='pip3 -q'
+APT='apt -qq -y'
 
-echo "* Installing go dependencies"
-  sudo apt install golang
-  mkdir -p $GOPATH
-  export PATH=$PATH:$GOPATH/bin
-
-echo "* Installing gnxi tools"
-  repo=github.com/google/gnxi
-  for tool in gnmi_{capabilities,get,set,target}; do
-    go get $repo/$tool
-    go install $repo/$tool
-  done
-
-echo "* Installing FAUCET as root"
-  sudo pip3 -q install --no-cache --upgrade faucet
+echo "* Installing python dependencies"
+  sudo $APT install python3-pip python3-setuptools python3-wheel
+  $PIP3 install protobuf grpcio grpcio-tools requests
 
 echo "* Done"

--- a/makefile
+++ b/makefile
@@ -12,7 +12,7 @@ extstubs = gnmi_ext_pb2.py gnmi_ext_pb2_grpc.py
 # GRPC/protobuf compiler command
 protoc = python3 -m grpc_tools.protoc --python_out=. --grpc_python_out=. -I.
 
-# Go path - required for gnmi_{set,get}
+# Set a default GOPATH if needed for gnmi_* tools used in test
 GOPATH ?= $(HOME)/go
 
 all: $(gnmistubs) $(extstubs)
@@ -34,7 +34,7 @@ $(extstubs): gnmi_ext.proto
 
 codecheck: $(gnmistubs) $(extstubs)
 	flake8 faucetagent.py agenttest.py
-	python3 -m pylint faucetagent.py agenttest.py
+	pylint faucetagent.py agenttest.py
 
 yapf:
 	yapf3 -i *py
@@ -43,5 +43,6 @@ clean:
 	rm -rf *proto *pb2*py *pyc *~ \#*\# testcerts *log __pycache__ *yaml
 
 test: all
+	@echo "* Using GOPATH=$(GOPATH)"
 	GOPATH=$(GOPATH) PATH=$(GOPATH)/bin:$(PATH) ./agenttest.py
 	grep faucet_config faucetagent.log || true

--- a/test-dependencies.sh
+++ b/test-dependencies.sh
@@ -1,14 +1,47 @@
 #!/bin/bash
+set -e
 
-echo "* Installing testing (make {test,codecheck}) dependencies"
-  pip3 -q install flake8 pylint
+PIP3='sudo pip3 -q install --no-cache'
+APT='sudo apt -qq -y install'
+
+echo "* Installing go dependencies"
+  $APT golang
+
+echo "* Checking for go version 1.7 or later"
+  goversion=$(go version | awk '{print $3;}')
+  latest=`printf "go1.7\n%s" $goversion | sort -V | tail -n1`
+  if [ "$latest" != "$goversion" ]; then
+      echo "gnxi requires go version >= go1.7 (found $goversion)"
+      exit 1
+  fi
+
+echo "* Installing gnxi tools"
+  if [ "$GOPATH" == "" ]; then
+    export GOPATH=$HOME/go
+    echo "* GOPATH not set - using $GOPATH"
+  fi
+  mkdir -p $GOPATH
+  export PATH=$GOPATH/bin:$PATH
+  repo=github.com/google/gnxi
+  for tool in gnmi_{capabilities,get,set,target}; do
+    go get $repo/$tool
+    go install $repo/$tool
+  done
+
+echo "* Installing python dependencies"
+  $PIP3 flake8 pylint protobuf grpcio grpcio-tools requests
+
+echo "* Installing latest faucet"
+  $PIP3 --upgrade git+https://github.com/faucetsdn/faucet
+
+echo "* Installing openvswitch-switch, gcc and latest mininet"
+  $APT openvswitch-switch
+  sudo service openvswitch-switch start
   TMPDIR=$(mktemp -d) && pushd $TMPDIR
   git clone https://github.com/mininet/mininet
-  pip3 -q install ./mininet
-  sudo pip3 -q install ./mininet
-  cd mininet && sudo make install-mnexec
+  cd mininet
+  sudo make install-mnexec
+  $PIP3 .
   popd && sudo rm -rf $TMPDIR
-  sudo apt install openvswitch-switch
-  sudo service openvswitch-switch start
 
 echo "* Done"

